syntax = "proto3";

message RequestGetVersion {
}

message ResponseGetVersion {
  string version = 1;
}

message RequestGetMasterFingerprint {
}

message ResponseGetMasterFingerprint {
  uint32 fingerprint = 1;
}

message RequestGetExtendedPubkey {
  bool display = 1;
  repeated uint32 bip32_path = 2;
}

message ResponseGetExtendedPubkey {
  string pubkey = 1;
}

message RequestRegisterWallet {
  string name = 1;
  string descriptor_template = 2;
  repeated string keys_info = 3;
}

message ResponseRegisterWallet {
  bytes wallet_id = 1;
  bytes wallet_hmac = 2;
}

message RequestGetWalletAddress {
  bool display = 1;
  string name = 2;
  string descriptor_template = 3;
  repeated string keys_info = 4;
  optional bytes wallet_hmac = 5;
  bool change = 6;
  uint32 address_index = 7;
}

message ResponseGetWalletAddress {
  string address = 1;
}

message RequestSignPsbt {
  bytes psbt = 1;
  string name = 2;
  string descriptor_template = 3;
  repeated string keys_info = 4;
  optional bytes wallet_hmac = 5;
}

message PartialSignature {
  bytes signature = 1;
  bytes public_key = 2;
  bytes leaf_hash = 3; // Optional. If it's not set, it will have the default value (empty).
}

message ResponseSignPsbt {
  repeated PartialSignature partial_signatures = 1;
}


message ResponseError {
  string error_msg = 1;
}

message Request {
  oneof request {
    RequestGetVersion get_version = 1;
    RequestGetMasterFingerprint get_master_fingerprint = 2;
    RequestGetExtendedPubkey get_extended_pubkey = 3;
    RequestRegisterWallet register_wallet = 4;
    RequestGetWalletAddress get_wallet_address = 5;
    RequestSignPsbt sign_psbt = 6;
  }
}

message Response {
  oneof response {
    ResponseGetVersion get_version = 1;
    ResponseGetMasterFingerprint get_master_fingerprint = 2;
    ResponseGetExtendedPubkey get_extended_pubkey = 3;
    ResponseRegisterWallet register_wallet = 4;
    ResponseGetWalletAddress get_wallet_address = 5;
    ResponseSignPsbt sign_psbt = 6;
    ResponseError error = 7;
  }
}
